app/models/product.rb

```diff
@@ -1,12 +1,34 @@
+class UrlValidator < ActiveModel::EachValidator
+  def validate_each(record, attr, value)
+    unless value =~ %r{\.(gif|jpg|png)\z}i
+      record.errors.add attr, 'must be a URL for GIF, JPG or PNG images'
+    end
+  end
+end
+
+class PriceValidator < ActiveModel::Validator
+  def validate(record)
+    if record[:price] < record[:discount_price]
+      record.errors.add :price, "should be greater than #{record[:discount_price]}"
+    end
+  end
+end
+
 class Product < ApplicationRecord
 
+  validates :permalink, uniqueness:true
+  validates :permalink, format: {
+    with: %r{\A[A-Za-z0-9-]+\Z},
```

Comment:
- regex should be constant
- do we need both A-Z and a-z? 
- allow_nil?

Reply:
done